tables: implement `password_policy` table for macOS

sharvilshah opened the pull request:
Uses OpenDirectory and output is similar to apple's built-in `pwpolicy` command. 

```
osquery> select * from password_policy;
+-------------------------------------------------------------------------+----------------------------------------------------------------+-----------------------------------------------------------+
| policy_identifier                                                       | policy_attribute                                               | policy_description                                        |
+-------------------------------------------------------------------------+----------------------------------------------------------------+-----------------------------------------------------------+
| ProfilePayload:9c546b0a-53f7-4f33-8bc6-84ec075817f5:minLength           | policyAttributePassword matches '.{10,}'                       | Contain at least 10 characters.                           |
| com.apple.defaultpasswordpolicy.fde                                     | policyAttributePassword matches '.{4,}+'                       | Enter a password that is four characters or more.         |
| ProfilePayload:9c546b0a-53f7-4f33-8bc6-84ec075817f5:requireAlphanumeric | policyAttributePassword matches '^(?=.*[0-9])(?=.*[a-zA-Z]).+' | Contain at least one number and one alphabetic character. |
+-------------------------------------------------------------------------+----------------------------------------------------------------+-----------------------------------------------------------+
```

This covers default policy (`com.apple.defaultpasswordpolicy.fde` in the example above) as well as ones pushed by MDM profiles.

directionless commented on specs/darwin/password_policy.table:
```diff
@@ -0,0 +1,11 @@
+table_name("password_policy")
```
At least on my mac, this is documented both as having global policies, as well as per-user policies. Any thoughts about how to handle that?

sharvilshah replied:
Yes, the reason I used `ODNodeCopyAccountPolicies` API call is to get both global and per-user policies. I am not sure on other platform, but windows has `NetUserModalsGet` in it's win32 API, from their description:


>The NetUserModalsGet function retrieves global information for all users and global groups in the security database, which is the security accounts manager (SAM) database or, in the case of domain controllers, the Active Directory.

From: https://docs.microsoft.com/en-us/windows/win32/api/lmaccess/nf-lmaccess-netusermodalsget?redirectedfrom=MSDN

I am guessing that should cover it? I am not sure on Linux

sharvilshah replied:
On macOS, the concept of `user` policies and `global` policies has been deprecated for a while now. From the man page:

>  Account Policies
     Account policies are the replacement for the deprecated legacy global and user policies.  Account policies are specified as a dictionary containing three keys, one key for each policy category.  Note that the dictionary is
     not required to contain all of the policy categories
     
The `policy_identifier` is something that a user can set (either manually or via MDM). Apple supplied policy has identifiers that start with `com.apple.`, and currently Apple only supplies one policy by default -- where if filevault gets enabled, the password has to be 4 characters or more.

For other platforms, if there is a clear user/global distinction, I guess one can use the extended schema for the table.

directionless commented on osquery/utils/conversions/darwin/cfdictionary.cpp:
```diff
@@ -0,0 +1,43 @@
+
+namespace osquery {
+
+std::string getPropertiesFromDictionary(const CFDictionaryRef& dict,
```
I'm fine either way, but does this make sense to take a pointer receiver and return a status? (I have no idea what our best practice is. But it's what it looks like `CFDictionaryGetValueIfPresent` uses. Though I guess that's CF, and not osquery.)

Either way, do we want tests?

sharvilshah replied:
We want const references and not a pointer receiver. This also matches our current style. Also added tests, thanks!

directionless commented on osquery/tables/system/darwin/password_policy.cpp:
```diff
@@ -0,0 +1,90 @@
+  auto node = ODNodeCreateWithNodeType(
+      kCFAllocatorDefault, kODSessionDefault, kODNodeTypeLocalNodes, &error);
+  if (node == nullptr) {
+    VLOG(1) << "Error creating an OpenDirectory node";
```
How often does this occur? It sounds like enough of an error that it should be more than verbose. Please also include the table name

directionless commented on osquery/tables/system/darwin/password_policy.cpp:
```diff
@@ -0,0 +1,90 @@
+  // have been deprecated
+  auto policies = ODNodeCopyAccountPolicies(node, &error);
+  if (policies == nullptr) {
+    VLOG(1) << "Error getting account policies";
```
same

directionless commented on osquery/tables/system/darwin/password_policy.cpp:
```diff
@@ -0,0 +1,90 @@
+
+  auto count = CFDictionaryGetCount(policies);
+  if (count == 0) {
+    VLOG(1) << "Empty account policies for the node";
```
This one sounds like verbose? But table name

directionless commented on osquery/tables/system/darwin/password_policy.cpp:
```diff
@@ -0,0 +1,90 @@
+
+  if (!CFDictionaryContainsKey(policies,
+                               CFSTR("policyCategoryPasswordContent"))) {
+    VLOG(1) << "Account policy does not contain password content";
```
same

directionless commented on osquery/tables/system/darwin/password_policy.cpp:
```diff
@@ -0,0 +1,90 @@
+        (CFDictionaryRef)CFArrayGetValueAtIndex((CFArrayRef)content, i),
+        "policyIdentifier");
+
+    auto dict = CFArrayGetValueAtIndex((CFArrayRef)content, i);
```
Should this be on line 65, and used by `policy_attribute` and `policy_identifier` ?

sharvilshah replied:
No, because the policy dictionary is a meta dictionary -- and the value of `policyCategoryPasswordContent` is an array of dictionaries, with attribute and description. 

Added a block comment to document the structure.

directionless commented on osquery/tables/system/darwin/password_policy.cpp:
```diff
@@ -0,0 +1,90 @@
+    results.push_back(r);
+  }
+
+  CFRelease(policies);
```
Is this being leaked on the early return paths? Does `scope_guard` make sense here?

sharvilshah replied:
fixed

sharvilshah commented:
Added one more column to make it easy to show the relevant data:

```
osquery> select * from password_policy;
+-------------------------------------------------------------------------+---------------------+----------------------------------------------------------------+-----------------------------------------------------------+
| policy_identifier                                                       | policy_attribute    | policy_content                                                 | policy_description                                        |
+-------------------------------------------------------------------------+---------------------+----------------------------------------------------------------+-----------------------------------------------------------+
| ProfilePayload:9c546b0a-53f7-4f33-8bc6-84ec075817f5:minLength           | minLength           | policyAttributePassword matches '.{10,}'                       | Contain at least 10 characters.                           |
| com.apple.defaultpasswordpolicy.fde                                     |                     | policyAttributePassword matches '.{4,}+'                       | Enter a password that is four characters or more.         |
| ProfilePayload:9c546b0a-53f7-4f33-8bc6-84ec075817f5:requireAlphanumeric | requireAlphanumeric | policyAttributePassword matches '^(?=.*[0-9])(?=.*[a-zA-Z]).+' | Contain at least one number and one alphabetic character. |
+-------------------------------------------------------------------------+---------------------+----------------------------------------------------------------+-----------------------------------------------------------+
```
 Final review @directionless?

sharvilshah commented:
Hey @directionless,

Did a little bit more digging and testing, there is another obscure-ish API to grab account policies for the user, in case there is a per-user policy. I have implemented that too, and I added a `uid` column when that's the case.

```
squery> select * from password_policy;
+-----+-------------------------------------------------------------------------+----------------------------------------------------------------+-----------------------------------------------------------+
| uid | policy_identifier                                                       | policy_content                                                 | policy_description                                        |
+-----+-------------------------------------------------------------------------+----------------------------------------------------------------+-----------------------------------------------------------+
|     | ProfilePayload:9c546b0a-53f7-4f33-8bc6-84ec075817f5:minLength           | policyAttributePassword matches '.{10,}'                       | Contain at least 10 characters.                           |
|     | com.apple.defaultpasswordpolicy.fde                                     | policyAttributePassword matches '.{4,}+'                       | Enter a password that is four characters or more.         |
|     | ProfilePayload:9c546b0a-53f7-4f33-8bc6-84ec075817f5:requireAlphanumeric | policyAttributePassword matches '^(?=.*[0-9])(?=.*[a-zA-Z]).+' | Contain at least one number and one alphabetic character. |
| 502 | com.apple.policy.legacy.minChars                                        | policyAttributePassword matches '.{12,}?'                      | Must be a minimum of 12 characters in length              |
+-----+-------------------------------------------------------------------------+----------------------------------------------------------------+-----------------------------------------------------------+
```
In this example above, I created a new user, and set that user's account policy to have minimum length of the password to be 12 characters.

sharvilshah commented on osquery/tables/system/darwin/password_policy.cpp:
```diff
@@ -0,0 +1,133 @@
+  }
+
+  // iterate over users, and get policy for the user
+  auto users = SQL::selectAllFrom("users");
```
Yep! Updated

zwass commented on osquery/tables/system/darwin/password_policy.cpp:
```diff
@@ -0,0 +1,145 @@
+  for (const auto& uid : uids) {
+    auto uid_string = CFStringCreateWithCString(
+        kCFAllocatorDefault, uid.c_str(), CFStringGetSystemEncoding());
+    auto query = ODQueryCreateWithNode(kCFAllocatorDefault,
+                                       node,
+                                       (CFTypeRef)kODRecordTypeUsers,
+                                       kODAttributeTypeUniqueID,
+                                       kODMatchEqualTo,
+                                       uid_string,
+                                       nullptr,
+                                       0,
+                                       nullptr);
+    auto records = ODQueryCopyResults(query, false, nullptr);
```
Do these need to be released?

zwass replied:
It's not clear exactly which documentation applies to these opendirectory functions, but it seems like they probably do if these docs apply https://developer.apple.com/library/archive/documentation/CoreFoundation/Conceptual/CFMemoryMgmt/Concepts/Ownership.html#//apple_ref/doc/uid/20001148-103029

sharvilshah replied:
@zwass Ah, thanks! `query` doesn't need to be released, but `records` do, thanks for catching this! Updated

zwass reviewed (dismissed):
LGTM unless those objects do need to be released.

zwass reviewed (approved):
Sweet!

sharvilshah commented:
<img width="1787" alt="Screen Shot 2022-06-30 at 11 41 18 AM" src="https://user-images.githubusercontent.com/819943/176755034-8e2176c6-4de3-422f-bf27-653a97a035e4.png">

Hooked it up to the profiler, and no leaky objects are detected for `password_policy` queries!
